devito/finite_differences/differentiable.py

```diff
@@ -645,6 +645,39 @@ def __str__(self):
     __repr__ = __str__
 
 
+class ComplexPart(Differentiable, sympy.core.function.Application):
+    """Abstract class for `Re` or `Im` of an expression"""
+
+    def __new__(cls, *args, **kwargs):
+        if len(args) != 1:
+            raise ValueError(f"{cls.__name__} is constructed with exactly one arg;"
+                             f" {len(args)} were supplied.")
+
+        # Diffify any Add, Mul, etc which might be in the expression
+        new_args = (diffify(args[0]),)
```

Comment:
This should not be needed,  `diffify` is not meant for corner cases induced by sympy's internal hardcoding of some classes like Add and Mul in some places. `diffify` should technically only be used with the compiler pass not for user level objects.

Reply:
It's needed here to guarantee that the arg has a dtype attribute, otherwise the CSE can't find the correct dtype and reverts back to float32, which results in values getting cast where they shouldn't be

Reply:
You have an MFE for it? If `diffify` was needed for  a new `Differentiable` object it would be a huge problem throughout.

Reply:
It is a new behaviour specific to `ComplexPart` as it changes the type from complex to real, which was previously not possible in Devito